File: Android/src/androidMain/kotlin/com/taskbridge/android/repository/NavigationRepository.kt
package com.taskbridge.android.repository

import com.taskbridge.core.models.navigation.AppTab
import com.taskbridge.core.models.navigation.NavigationDestination
import com.taskbridge.core.models.navigation.NavigationDestinationMessage
import com.taskbridge.core.models.navigation.NavigationOverlay
import com.taskbridge.core.models.navigation.NavigationPath
import kotlinx.coroutines.flow.Flow

sealed interface NavigationDestinationMessageScopeId {
    val scopeId: String

    data object TasksRoot : NavigationDestinationMessageScopeId {
        override val scopeId: String = "tasksRoot"
    }

    data class TaskDetails(
        val parentId: String
    ) : NavigationDestinationMessageScopeId {
        override val scopeId: String = "taskDetails:$parentId"
    }

    data object RemindersRoot : NavigationDestinationMessageScopeId {
        override val scopeId: String = "remindersRoot"
    }
}

interface NavigationRepository {
    val activePath: Flow<NavigationPath?>
    val currentTab: Flow<AppTab>
    val overlay: Flow<NavigationOverlay?>

    suspend fun selectTab(tab: AppTab)
    suspend fun pushDestination(destination: NavigationDestination)
    suspend fun popDestination()
    suspend fun pullToRoot(tab: AppTab)
    suspend fun fetchActivePath(): NavigationPath?
    suspend fun fetchCurrentTab(): AppTab
    suspend fun setNavigationDestinationMessage(message: NavigationDestinationMessage?)
    suspend fun consumeNavigationDestinationMessage(scopeId: String): NavigationDestinationMessage?
}

File: iOS/TaskBridge/Repository/NavigationRepository.swift
import Foundation
import TaskBridgeCore
import Combine

public enum NavigationDestinationMessageScopeId: Equatable {
    case tasksRoot
    case taskDetails(parentId: String)
    case remindersRoot

    public var scopeId: String {
        switch self {
        case .tasksRoot:
            return "tasksRoot"
        case let .taskDetails(parentId):
            return "taskDetails:\(parentId)"
        case .remindersRoot:
            return "remindersRoot"
        }
    }
}

@MainActor
public protocol NavigationRepository {
    var activePath: AsyncStream<NavigationPath?> { get }
    var currentTab: AsyncStream<AppTab> { get }
    var overlay: AsyncStream<NavigationOverlay?> { get }

    func selectTab(tab: AppTab) async throws
    func pushDestination(_ destination: NavigationDestination) async throws
    func popDestination() async throws
    func pullToRoot(tab: AppTab) async throws
    func fetchActivePath() async throws -> NavigationPath?
    func fetchCurrentTab() async throws -> AppTab
    func setNavigationDestinationMessage(_ message: NavigationDestinationMessage?) async throws
    func consumeNavigationDestinationMessage(scopeId: String) async throws -> NavigationDestinationMessage?
}
